xarray/core/dataarray.py

```diff
@@ -140,7 +131,9 @@ def _check_coords_dims(shape, coords, dim):
 
 
 def _infer_coords_and_dims(
-    shape, coords, dims
+    shape: tuple[int, ...],
+    coords: Sequence[Sequence[Any] | pd.Index | DataArray] | Mapping[Any, Any] | None,
```

Comment:
Very minor, but unless we're explicitly stating that these can contain `Any`, I would vote to keep them small:

```suggestion
    coords: Sequence[Sequence | pd.Index | DataArray] | Mapping | None,
```